### PR TITLE
chore(infrastructure): update webpack v4.x

### PR DESCRIPTION
**Pull request by moog16**
fixes #4847 

webpack3 (master branch):
- running `npm run dist`:
  - 2:22
  - 2:30
  - 2:17
  - average: 143s

~~webpack4 (with 2 minified file build):~~
~~- running `npm run dist`:~~
  ~~- 1:09~~
  ~~- 1:09~~
  ~~- 1:12~~

webpack4 (with minified and non-minified builds):
- running `npm run dist`:
  - 1:08
  - 1:08
  - 1:11
  - average: 71.7s

**Build time decreased by 49.14%**

### CODE REVIEW

**Comment by abhiomkar** [comment]
Why would be remove *.min.js files?

**Comment by abhiomkar** [comment]
Did you see any different in build time compared to Webpack 3?

**Comment by moog16** [comment]
by default they the bundles are all minified. So there's no need to have both as they are the same.

**Comment by abhiomkar** [comment]
We need to continue publishing both minified and un-minified versions like before. Un-minified version is useful for development / debugging.

**Comment by moog16** [comment]
We can provide maps instead of un-minified (which we already actually do). I don't think un-minified is totally necessary.

**Comment by abhiomkar** [comment]
We shouldn't break existing publish structure as part of this upgrade, IMO.

I'm not expert in this area, but it is fairly common practice to publish minified & un-minified.

**Comment by moog16** [comment]
Fair point! I'll make it exactly as it was.

File: scripts/build/environment.js [hunk 1]
```diff
@@ -27,6 +27,8 @@
 
 'use strict';
 
+// TODO: remove this class. this is only used for webpack.
```
Comment (moog16) [inline scripts/build/environment.js]: I can add an issue for this, but I didn't wanna do this without chatting with the group about it. This seems to be unnecessary for webpack to work.

**Comment by moog16** [comment]
@abhiomkar ready to go again

File: scripts/build/environment.js (hunk 1 above)
Comment (abhiomkar) [reply scripts/build/environment.js]: yah, we can remove this in upcoming PR.

File: scripts/cp-pkgs.js [hunk 2]
```diff
@@ -110,20 +109,25 @@ async function cpAsset(asset) {
...
-    const destBasename = isAllInOne ? packageDirectory : `mdc.${toCamelCase(packageDirectory.replace(/^mdc-/, ''))}`;
-    const destFilename = path.join(packagePath, 'dist', `${destBasename}.d.ts`);
-
-    console.log(`Writing UMD declarations in ${destFilename.replace(process.cwd() + '/', '')}`);
-    dts.bundle({
-      name,
-      main,
-      out: destFilename,
+    const main = path.join(PACKAGES_DIRECTORY, packageDirectory, './index.d.ts');
+    fs.access(main, fs.constants.F_OK, (err) => {
```
Comment (abhiomkar) [inline scripts/cp-pkgs.js]: nit: error*

File: scripts/cp-pkgs.js [hunk 3]
```diff
@@ -110,20 +109,25 @@ async function cpAsset(asset) {
...
+      if (!err) {
```
Comment (abhiomkar) [inline scripts/cp-pkgs.js]:
```ts
if (error) {
  return;
}
```

Returns early if error exist. This avoids nested block and makes it easy to read.

File: scripts/build/environment.js (hunk 1 above)
Comment (moog16) [reply scripts/build/environment.js]: created issue https://github.com/material-components/material-components-web/issues/4863

**Pull request merged**